acvm-repo/acvm/src/compiler/optimizers/redundant_range.rs

```diff
@@ -139,20 +164,64 @@ impl<F: AcirField> RangeOptimizer<F> {
                 continue;
             };
 
+            let next_side_effect = next_side_effects[idx];
```

Comment:
This feels a little roundabout with `collect_side_effects`. Could we not pass `brillig_side_effects: &BTreeMap<BrilligFunctionId, bool>,` to `collect_ranges` match on an `Opcode::BrilligCall` here and return `next_side_effect` directly from the match based upon the `BrilligFunctionId`. We could then avoid enumerating this loop.

Reply:
The interaction between `collect_side_effects` and `collect_ranges` certainly evolved. Originally `collect_side_effects` tracked inputs and asserts, and I didn't want to burden `collect_ranges` with that. 

However, I don't see immediately how we can avoid enumerating this loop. Here we go forward in the opcodes, while `collect_side_effects` provides information about higher indexes, opcodes this enumeration hasn't seen yet, to answer whether there will be things before the next thing happens. 

`collect_ranges` could build up two data structures, but I'm not sure it would be easier to understand that way.

Happy to revisit this, or if you can simplify it. Just want to get it right, and still need to look at the recursion in the Brillig opcodes 😬

Reply:
I removed `collect_side_effects`. Instead `next_side_effect` is maintained during `replace_redundant_ranges`, which now goes backwards. Also simplified the switch points to just be a set of indexes where we might want range constraints to be kept.